firebaseおよびapiClientの作成

toshinori-m opened the pull request:
### 変更内容
- Firebase 認証の導入
  - firebase 設定と auth のセットアップを実装
  - Google サインイン機能を追加
- 認証コンテキスト (AuthContext.tsx, useAuthContext.ts) の作成
  - AuthContext を作成し、ユーザー情報をグローバルに管理
  - useAuthContext.ts で useAuth / useSetAuth フックを定義し、認証情報を簡単に取得・更新できるように
- API クライアント (apiClient.ts) の作成
  - バックエンド API との通信を簡単にするためのクライアントを作成
- SignupPage.tsx の作成
  - Google サインインを実装し、認証後に API にユーザー情報を送信する処理を追加

### 動作確認方法
1. `http://localhost:5173/signup` にアクセス
2. `Googleアカウントで登録`ボタンをクリックし、下を確認
   - Firebase にメールアドレスと UID が登録
   - Docker の PostgreSQL に UID、name、email が保存
3. 再度 `Googleアカウントで登録` ボタンをクリックし、既存ユーザーのエラーメッセージが表示されることを確認 
   - `このメールアドレスは既に登録されています。ログイン画面からGoogleアカウントでログインして下さい`

close https://github.com/toshinori-m/stay_connect/issues/165

toshinori-m committed:
firebaseおよびapiClientの作成

toshinori-m committed:
CORS設定の修正、エラーハンドリングの改善、ログメッセージ修正

hirosi1900day commented on frontend-react/src/types.ts:
```diff
@@ -1,6 +1,36 @@
+    }
+  }
+}
+
```
typeとinterfaceで分けているのには何か意図があったりしますか？？
ないのであれば統一すると良いかと思います

hirosi1900day commented on frontend-react/src/types.ts:
```diff
@@ -1,6 +1,36 @@
+// firebase のエラー型
+export interface FirebaseError {
+  code: string;
+}
```
FirebaseErrorは"firebase/app"からimportするではダメでしょうか？？
独自で定義するとライブラリー側がエラーの型を変えた時に急にアプリケーションが動かなくなることがあるので、ライブラリーにある型をできるだけ使用したほうが良いかと

hirosi1900day commented on frontend-react/src/types.ts:
```diff
@@ -1,6 +1,36 @@
+  name: string
+  email: string
+  uid: string
+}
```
この型はどこかで使ってます？？

hirosi1900day commented on frontend-react/src/pages/SignupPage.tsx:
```diff
@@ -1,16 +1,68 @@
+
   const handleRegisterClick = () => {
-    console.log("次のissueで作成予定！")
+    console.log("メールアドレス画面は次のissueで作成予定！")
```
以下みたいにするとTODOで検索かけるとまだ未実装の箇所をすぐに見つけ出すことができるので実務では良くやります
```ts
# TODO: 後続タスクで処理を追加する
console.log("メールアドレス画面は次のissueで作成予定！")
```

https://sqripts.com/2022/04/19/20423/

hirosi1900day commented on frontend-react/src/hooks/apiClient.ts:
多分hooksディレクトリーの処理ではなくlibとかに配置するべき処理かなと思ったのですがhooksにおいたのには理由がありますか？？(一理hookに書くという考え方もわからなくはない)

lib配下においてcontextに依存しない形にしてsetAuthHeaderで各コンポーネントで冗長ではあるがsetAuthHeaderを使ってuidを仕込む方がわかりやすかったりしないか？という意見です
hooksにした意図を聞いてみたいです！

```ts
import axios from "axios";

// axios インスタンスを作る
export const axiosInstance = axios.create({
  baseURL: "https://api.example.com",
  headers: {
    "Content-Type": "application/json",
  },
});

// `uid` を設定する関数
export function setAuthHeader(uid: string) {
  axiosInstance.defaults.headers.Authorization = `Bearer ${uid}`;
}
```

またはlibではなくhookで扱うとしたらdataとisLoading なども返すなど状態を返すようにするとhookに置く妥当性が高まる気がしました！

hirosi1900day commented on frontend-react/src/hooks/apiClient.ts:
```diff
@@ -0,0 +1,26 @@
+export const useApiClient = () => {
+  const { user } = useAuth()
+
+  useEffect(() => {
```
useEffectで取得する必要があるものでしょうか？？

hirosi1900day commented on frontend-react/src/hooks/apiClient.ts:
```diff
@@ -0,0 +1,26 @@
+
+const apiClient = axios.create({ baseURL: apiBaseUrl })
+
+export const useApiClient = () => {
```
userの認証有無を確認してheaderが内部で変わるようになってますが、どちらかというと使う側に認証しているかどうかを意識して使い分けた方が可読性的には良いのかなと思うので
例えばauthenticatedApiClientとunauthenticatedApiClientみたいな関数を作って分けて使用がわで使い分けてもらう方が良いかなと思いましたがいかがでしょうか？？

hirosi1900day commented on frontend-react/.gitignore:
```diff
@@ -22,3 +22,8 @@ dist-ssr
+# local env files
+.env.local
+.env.*.local
+.env
```
以下みたいにしてシンプルに書けないですかね　
```suggestion
.env*
!.env.example
```

hirosi1900day commented on frontend-react/src/context/AuthContext.ts:
```diff
@@ -0,0 +1,4 @@
+import { createContext } from "react"
+import { AuthContextType } from "@/types"
+
+export const AuthContext = createContext<AuthContextType | null>(null)
```
frontend-react/src/context/AuthProvider.tsx で管理した方が処理がまとまっておりみやすくないでしょうか？？

hirosi1900day commented on frontend-react/src/context/AuthProvider.tsx:
```diff
@@ -0,0 +1,22 @@
+  }, [])
+
+  return (
+    <AuthContext.Provider value={{ user, setUser }}>
```
これは発展課題なので、今回は対応しなくても良いですがset用のcontextとget用のcontextは分けると不要なサイレンダリングが防げてパーフォーマンスに良いです
https://zenn.dev/yuta_ura/articles/react-context-api

hirosi1900day replied:
react講座の117から120を見るとわかりやすいかと思います！

hirosi1900day commented on frontend-react/src/context/useAuth.ts:
```diff
@@ -0,0 +1,8 @@
+  const context = useContext(AuthContext)
+  if (!context) throw new Error("useAuth must be used within an AuthProvider")
+  return context
+}
```
分けると意味はあまりなさそうなのでfrontend-react/src/context/AuthProvider.tsxなどに集約した方が良いかなと思いましたがいかがでしょうか？？(高凝集と疎結合が原則)

https://qiita.com/kenji-kondo/items/bb25abd355b9fe660057

toshinori-m replied:
当初、AuthProvider.tsx、AuthContext.tsx、useAuth.ts を 1つのファイルに集約し、AuthProvider、AuthContext、useAuth を 1つのファイルからエクスポートするようにしていました。
エクスポート方法によってエラーが発生したため、最終的に分けることにしました。
下に、試したエクスポート方法とエラー内容を示します。
<方法1>
```
export { AuthProvider, AuthContext }
export default useAuth
```
「useAuthのエラー内容」
`Fast refresh only works when a file only exports components. Use a new file to share constants or functions between components.`

<方法2>
```
export default {
  AuthProvider,
  useAuth,
  AuthContext,
}
```
「AuthProviderのエラー内容」
`Fast refresh only works when a file only exports components. Move your component(s) to a separate file`

<方法3>
`export { AuthProvider, AuthContext, useAuth }`
「useAuthのエラー内容」
`Fast refresh only works when a file only exports components. Use a new file to share constants or functions between components.`

全てを集約するにはどのようにすればいいでしょう？改善方法を教えていただきますとありがたいです！

よろしくお願いします

hirosi1900day replied:
udemyのreact講座の117から120のようにアロー関数にして export constで展開する方法だとうまく行きませんでしたかね？？
https://www.udemy.com/course/react-complete-guide/learn/lecture/33043724?start=345#reviews

toshinori-m replied:
まだ未完成ですが
下のようにアロー関数にしてみました
[AuthProvider.tsx](https://github.com/toshinori-m/stay_connect/blob/388f21f179692083a646818dd62caaeb8ec4f2c1/frontend-react/src/context/AuthProvider.tsx)
すると下の注意が出ています
`Fast refresh only works when a file only exports components. Use a new file to share constants or functions between components.e`

アプリからgoogle認証を実行してみると正常に起動します
そうした場合、上の注意は気にする必要はないのでしょうか？

hirosi1900day replied:
ワーニングがでるのですね

ワーニングなので問題ないのはないですが気になるのでコンテキストの使用部分だけファイルを分けた方が良さそうですね🙏

ファイルを分ける場合はディレクリー構成でコンテキスト作成ファイルとコンテキストを使用するファイルの関連がわかりやすいようにしてみてください
https://qiita.com/kotoha__/items/a6ebd9e87a9f999cc681

hirosi1900day commented on frontend-react/src/firebase/firebase.ts:
初期設定を行う処理だと思うのでlibとかに配置するのが良かったりしませんか？？

hirosi1900day reviewed (commented):
まだ未完成だと思いますが、修正箇所が結構多そう and 完成した後に直すのも大変だと思うので一旦コメントしておきます🙏

hirosi1900day commented on frontend-react/src/types.ts:
```diff
@@ -1,6 +1,25 @@
+      error?: string
+    }
+  }
+}
```
APIのエラーレスポンスはサーバーごとに異なるため、`ApiError` のように抽象的すぎる名前で型を定義すると、チーム開発時に「どの場面でこの型を使えばよいのか」が分かりづらくなります。

もし、Railsで返すAPIのエラー型が **OpenAPI などで明確に定義されている** 場合は、`RailsApiError` のように **具体的な型名** にすることで、他の開発者が迷わずに使えるようになります。

また、Firebase のエラーについては **Firebase 固有のエラー型** を使って判定するようにしてください。

---

## 💡 補足
エラー型を統一する場合は、[RFC 9457](https://www.rfc-editor.org/rfc/rfc9457) に準拠した設計を採用し、Rails 側のエラーレスポンスもそれに沿う形で返すのが望ましいです。

web開発でちゃんと設計する際はRFCを見ることになるので、中身を全部覚える必要はないですが、基本方針ここの設計方針にしたがうと覚えておくといつか役に立つかもしれません

ちなみに超余談ですがapi設計は基本方針私は[API Improvement Proposals](https://google.aip.dev/)を見て設計することにしてます

hirosi1900day commented:
レビューリクエストが入ってますがもう再レビューしちゃっても良いですか？？

toshinori-m commented:
下の箇所がまだ未解決なので、レビューはしばらくお待ちください！
質問時にもボタンを押すものだと思っていました。

> 多分hooksディレクトリーの処理ではなくlibとかに配置するべき処理かなと思ったのですがhooksにおいたのには理由がありますか？？(一理hookに書くという考え方もわからなくはない)
> 
> lib配下においてcontextに依存しない形にしてsetAuthHeaderで各コンポーネントで冗長ではあるがsetAuthHeaderを使ってuidを仕込む方がわかりやすかったりしないか？という意見です
> hooksにした意図を聞いてみたいです！

よろしくお願いします

toshinori-m committed:
Firebase認証とAuthContextの実装修正

toshinori-m commented:
完成出来たように感じます
再度レビューをよろしくお願いします

toshinori-m committed:
.gitignore修正

hirosi1900day commented on backend/config/initializers/cors.rb:
```diff
@@ -7,7 +7,12 @@
+      ENV['BRANCH_ORIGIN'],
+      ENV['FRONTEND_ORIGIN'],
+      'http://localhost:81',
+      'http://localhost:5173'
```
```suggestion
      'http://localhost:81', # vue front用
      'http://localhost:5173' # react front用
```

readmeも修正しておくと良いかもしれないですね

hirosi1900day commented on frontend-react/src/context/AuthContext.tsx:
```diff
@@ -0,0 +1,27 @@
+import { onAuthStateChanged, User } from "firebase/auth"
+import { AuthUserContextType, AuthUserUpdateContextType } from "@/types"
+
+export const AuthContext = (() => createContext<AuthUserContextType | null>(null))()
+export const AuthUpdateContext = (() => createContext<AuthUserUpdateContextType | null>(null))()
```
providerの中でcontextが使用されていることをチェックするためにnull許容している認識であっていますかね？？
となるとnullよりもundefined許容にする方が適切かもしれないです

もし別の意図があればコメントください

toshinori-m replied:
このnullについて迷い「[undefinedとnullの違い](https://typescriptbook.jp/reference/values-types-variables/undefined-vs-null)」をみてnullにしていました
この記事から下のように捉えていました
- undefined : 値が未定義
- null : 意図的に値がない

・undefined を選んだ場合、「まだ値が設定されていないのか」「そもそも値が提供されていないのか」が分かりにくくなると感じました。

・Context の場合は 「意図的に null を初期値とすることで、値がまだ設定されていないことを明確にする」 ことが目的のように感じたのでnull を選択しました。

よろしくお願いします。

hirosi1900day replied:
考えがあってあえてnullにしているのであれはこの形でも大丈夫です！

とはいえ、私がこれまで関わってるプロジェクトなどでは基本undefinedが使われている印象なので今後複数のプロジェクトに関わったときに複数の考え形があるということは覚えておいても良いかもですね！（このあたりは考え形に違いがあっても問題ありません。）
https://www.carlrippon.com/react-context-with-typescript-p4/

hirosi1900day commented on frontend-react/src/lib/apiClient.ts:
```diff
@@ -0,0 +1,22 @@
+  }
+
+  return axiosInstance
+}
```
クライアント側でuidを指定するのはなく、useApiClient側にそのロジックを持つのであれば責務的には
ここだけはuseAuth を使っているカスタムフック なのでhooksディクレリーに置くべき処理ですかね！

また以下のようにuseMemoを使ってクライアント生成ができると不要なレンダリングにより、何度もインスタンスのセットが行われないので良いかもしれないですね！(動作確認はしてないのでイメージです)
```ts
export const useApiClient = () => {
  const { user } = useAuth()

  const apiClient = useMemo(() => createAxiosInstance(user?.uid), [user?.uid])

  return apiClient
}

```

hirosi1900day commented on frontend-react/src/pages/SignupPage.tsx:
```diff
@@ -1,16 +1,81 @@
+      }
+    
+      setError("予期しないエラーが発生しました。")
+    }
```
これは今回のPRで対応しなくても大丈夫ですが、実際の業務コードでは意識しておいてほしいポイントなので、別途 issue にしておいてもらえると助かります。

考慮が必要なケース:
「Firebase の登録には成功したが、Rails のユーザー登録で失敗した場合」

この場合、Rails にはユーザーが登録されていませんが、Firebase にはすでにアカウントが作成されている状態になります。そのため、次回同じユーザーが登録を試みたときに、Firebase 側で auth/credential-already-in-use エラーが発生する可能性があります。(このケースが発生してしまうとそのユーザーはもう一生このサイトに登録できなくなってします)

このようなケースでは、Rails の登録が失敗したタイミングで、Firebase 側のユーザーも削除する処理を追加する必要があります。（現在の実装では、異常系の考慮が足りていない可能性があります。）

また、setUser(result.user) を Firebase の認証直後、Rails のユーザー登録前に実行しているため、Rails 側でエラーが発生しても、認証済みとして処理されてしまいます。そのため、Rails の登録に失敗した場合は、setUser(null) に設定し直す必要があると思います。

toshinori-m replied:
下を作成しましたので、後日作業を行います
issue：Firebase登録成功後にRailsユーザー登録が失敗した場合の処理

hirosi1900day commented on frontend-react/src/types.ts:
```diff
@@ -1,6 +1,27 @@
+}
+export interface AuthUserUpdateContextType {
+  setUser: (user: FirebaseUser | null) => void
+}
```
この型ははfrontend-react/src/context/AuthContext.tsxでしか使わないと思うのでそちらのファイルに記載した方が可読性が良いかなと思います(そのファイルでしか使わないものはできるだけそのファイルに閉じた方が読み手は読みやすいかなと思います)

このファイルに書くものは複数のファイルで使用する and 適切なファイルが見つからない際にこちらに記載するイメージが良いかなと思います！

hirosi1900day reviewed (commented):
だいぶ良くなってきましたが
追加でコメントしました！(あと一息)

hirosi1900day commented:
[nits]コミットメッセージにwipが入っているのでレビュー依頼前にコミットを綺麗にしておいてください

toshinori-m committed:
API クライアントの責務整理、認証コンテキストの型の整理
- useApiClient 内で useMemo を使用し、不要な再レンダリングを防止
- 認証コンテキストの型を frontend-react/src/context/AuthContext.tsx へ移動
- Vue（localhost:81）と React（localhost:5173）フロントの URL コメントを追加

hirosi1900day reviewed (approved):
おつかれさまです！
